designs/2021-suppression-support/README.md

```diff
@@ -0,0 +1,302 @@
+2.	When the flag `--output-suppression` is used, the violations (problems) that disabled by inline comments will not be removed.
+3.	Add suppression info in the output of ESLint
+    1. For inSource suppressions, we generate suppression info where `kind` is `inSource` and `justification` is the description in the directive comments (words after two or more consecutive dashes mentioned above).
+    
+        For example:
+        
+        ```javascript
+        /* eslint-disable no-undef -- Justification example */
+        a = 1;
+        /* eslint-enable no-undef */
+        ```
+        ```javascript
+        console.log("foo"); // eslint-disable-line no-console -- Justification example
+        ```
+        ```javascript
+        // eslint-disable-next-line no-console -- Justification example
+        console.log("bar");
+        ```
+
+        Output:
+
+        ```json
+        {"kind": "inSource", "justification": "Justification example"}
+        ```
+    
+    2. For external suppressions, we generate suppression info where `kind` is `external` and `justification` is a certain sentence such as "Globally disabled".
+
+        For example:
+
+        ```json
+        rules: {
+          "no-undef": "off"
+        }
+        ```
+
+        Output:
+
+        ```json
+        {"kind": "external", "justification": "Globally disabled"}
+        ```
+
+### General Design
+
+We are proposing to add a new CLI option `--output-suppression` to output suppressions without prompting disabled errors/warnings.
+
+![Design diagram](design_diagram.png)
+
+Input:
+
+```
+.\node_modules\.bin\eslint.cmd 1.js --output-suppression -f @microsoft/eslint-formatter-sarif -o 1.sarif
+```
+
+Output:
+
+![Suppression example](suppression_example.png)
+
+### Add new CLI option `--output-suppression`
+
+We propose to add a new CLI option `--output-suppression` for the new feature. Without the option, ESLint will work as it is currently, and `suppressions` property should not exist in the output. With this option, ESLint is expected to export all violations, including suppressed ones. `suppressions` should be _empty_ for non-suppressed violations, while `suppressions` should be a list for suppressed ones, according to [suppressions property](https://docs.oasis-open.org/sarif/sarif/v2.0/csprd02/sarif-v2.0-csprd02.html#_Toc10127852).
+
+### Parse description in comments as `justification`
+
+We propose not to ignore the text after two or more dashes in comments and considered it as the justification of a suppression. The modification should be in `getDirectiveComments`.
+
+### Add suppressions attribute in `LintMessage`
+
+`LintMessage` is the type that finally output to the formatter. Currently it contains `ruleId`, `severity`, `message`, etc. What we desire is a `suppressions` attribute in `LintMessage`. It should be a list that contains a suppression entry (`{kind: "inSource", justification: "Fake justification message."}`).
```

Comment:
Should suppressions be there even when the command line option isn’t used?

Reply:
No. We would like to keep the current behavior of ESLint. The property `suppressions` exists only when the option `--output-suppression` used.